tests/unit/classifier/svm/LibLinear_unittest.cc

```diff
@@ -6,1320 +6,480 @@
+			labels[i/2] = (i < data.num_cols/2) ? 1.0 : -1.0;
+		}
+		/*
+			Now fill in train_feats, test_feats, and ground truth
+		*/
 
-	generate_data_l2(train_feats, test_feats, ground_truth);
+		train_feats = (CDenseFeatures<float64_t>*)features.copy_subset(train_idx);
+		test_feats =  (CDenseFeatures<float64_t>*)features.copy_subset(test_idx);
+		SG_REF(train_feats);
+		SG_REF(test_feats);
 
-	CLibLinear* ll = new CLibLinear();
+		if (type == "L1")
+		{
+			//transpose
+			SGMatrix<float64_t> train_matrix = train_feats->get_feature_matrix();
+			SGMatrix<float64_t>::transpose_matrix(train_matrix.matrix,
+					train_matrix.num_rows, train_matrix.num_cols);
 
-	CBinaryLabels* pred = NULL;
-	float64_t liblin_accuracy;
-	CContingencyTableEvaluation* eval = new CContingencyTableEvaluation();
+			SG_UNREF(train_feats);
 
-	ll->set_bias_enabled(false);
-	ll->set_features(train_feats);
-	ll->set_labels(ground_truth);
+			train_feats = new CDenseFeatures<float64_t>(train_matrix);
 
-	ll->set_liblinear_solver_type(liblinear_solver_type);
-	ll->train();
-	pred = ll->apply_binary(test_feats);
-	SG_REF(pred);
-	liblin_accuracy = eval->evaluate(pred, ground_truth);
+			ground_truth = new CBinaryLabels(labels);
 
-	EXPECT_NEAR(liblin_accuracy, 1.0, 1e-6);
+			/*
+				Now we ref ground_truth[l1_index], train_feats[l1_index]  ( test_feats[l1_index] was referenced in copy_subset)
+			*/
+			SG_REF(ground_truth);
+			SG_REF(train_feats);
+		}
+		else
+		{
+				//type=="L2", no need to transpose
+				ground_truth = new CBinaryLabels(labels);
+				SG_REF(ground_truth);
+		}
+	}
+	void generate_data_simple(std::string type) //Type either "L1_SIMPLE" or "L2_SIMPLE"
+	{
+			int32_t rows, cols;
+			if (type=="L1_SIMPLE")
+			{
+				rows = 10;
+				cols = 2;
+			}
+			else
+			{
+				//type == L2_SIMPLE
+				rows = 2;
+				cols = 10;
+			}
+			SGMatrix<float64_t> train_data(rows, cols);
+			SGMatrix<float64_t> test_data(2, 10); //Always 2x10, doesn't matter l1 or l2
+
+			/*
+				We have to transpose the data if its l2. If it is l1, then leave it as it is (Since this is the data of l1 originally)
+			*/
+			index_t num_samples = 10;
+			std::vector<int32_t> x{0,0,1,1,2,2,3,3,4,4,5,5,6,6,7,7,8,8,9,9};
+			std::vector<int32_t> y{0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1};
+			std::vector<int32_t> z{-1,1,-1,0,-1,-1,-2,0,0,-1,2,2,3,1,3,2,3,3,4,2};
+
+			for(index_t i=0; i<num_samples*2; ++i){
+				if (type=="L1_SIMPLE")
+						train_data(x[i],y[i])=z[i];
+				else
+						train_data(y[i], x[i])=z[i]; //transpose
+				test_data(y[i], x[i])=z[i];
+			}
+			/*
+				The remainder is left regardless of weather this is for l1 or l2
+			*/
+			train_feats = new CDenseFeatures<float64_t>(train_data);
+			test_feats = new CDenseFeatures<float64_t>(test_data);
+
+			SGVector<float64_t> labels(num_samples);
+			for (index_t i = 0; i < num_samples; ++i)
+			{
+				labels[i] = (i < num_samples/2) ? 1.0 : -1.0;
+			}
+
+			ground_truth = new CBinaryLabels(labels);
+
+			SG_REF(ground_truth);
```

Comment:
all instances you are creating yourself could be created with `some` and then you dont need to REF and UNREF them.

Reply:
I did use some in the beginning but it leads to several problems since some function calls that instantiate elements do a SG_REF with new, so in the end, when I UNREF some of them, sometimes they don't need SG_UNREF (because they used some) and some need SG_UNREF because they used new. This lead to a lot of segmentation errors so I just replaced all with new so that I keep track of what memory i need to free.

Reply:
ah yes I know this problem. leave it for now then